src/ucp/core/ucp_mm.c

```diff
@@ -178,20 +183,107 @@ static ucs_status_t ucp_mem_alloc(ucp_context_h context, size_t length,
 }
 
 
+static inline unsigned uct_flags(ucp_mem_map_params_t *params)
+{
+    unsigned flags = 0;
+
+    if (params && (params->field_mask & UCP_MEM_MAP_PARAM_FIELD_FLAGS)) {
+        if (params->flags & UCP_MEM_MAP_NONBLOCK) {
+            flags |= UCT_MD_MEM_FLAG_NONBLOCK;
+        }
+
+        if (params->flags & UCP_MEM_MAP_FIXED) {
+            flags |= UCT_MD_MEM_FLAG_FIXED;
+        }
+    }
+
+    return flags;
+}
+
+/* Matrix of behavior
+ * |-------------------------------------------------------------------------------------------|
+ * | parameter |                              value                                            |
+ * |-----------|-------------------------------------------------------------------------------|
+ * | ALLOCATE  |   0   |     1    |   0   |  0  |   1   |      1        |  0  |     1          |
+ * | FIXED     |   0   |     0    |   1   |  0  |   1   |      0        |  1  |     1          |
+ * | addr      |   0   |     0    |   0   |  1  |   0   |      1        |  1  |     1          |
+ * |-----------|-------|----------|-------|-----|-------|---------------|-----|----------------|
+ * | result    | error | allocate | error | reg | error | allocate/hint | reg | allocate/fixed |
+ * |-------------------------------------------------------------------------------------------|
+ */
+static inline ucs_status_t ucp_mem_map_check_and_adjust_params(ucp_mem_map_params_t *params)
+{
+    if (!params) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    if (!(params->field_mask & UCP_MEM_MAP_PARAM_FIELD_LENGTH)) {
+        ucs_error("The length value for mapping memory isn't set: %s",
+                  ucs_status_string(UCS_ERR_INVALID_PARAM));
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    /* First of all, define all fields */
+    if (!(params->field_mask & UCP_MEM_MAP_PARAM_FIELD_ADDRESS) ||
+        ((params->field_mask & UCP_MEM_MAP_PARAM_FIELD_ADDRESS) &&
+         (params->address == NULL))) {
+        params->field_mask |= UCP_MEM_MAP_PARAM_FIELD_ADDRESS;
+        params->address = NULL;
+    }
+
+    if (!(params->field_mask & UCP_MEM_MAP_PARAM_FIELD_FLAGS)) {
+        params->field_mask |= UCP_MEM_MAP_PARAM_FIELD_FLAGS;
+        params->flags = 0;
+    }
+
+    if ((params->flags & UCP_MEM_MAP_FIXED) &&
+        !ucs_is_page_aligned(params->address)) {
+        ucs_error("UCP_MEM_MAP_FIXED flag requires page aligned address");
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    /* Now, lets check all erroneous cases from the matrix */
+    if (params->address == NULL) {
+        if (params->flags == 0) {
```

Comment:
these checks can be reduced to smth like this:

if ((params->flags & UCP_MEM_MAP_FIXED) ||
!(params->flags & UCP_MEM_MAP_ALLOCATE)) {
        return UCS_ERR_INVALID_PARAM;
}